// virtual-device-sdk/lib/src/ConsolePrinter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var chalk = require("chalk");
var path = require("path");
var TEST_NAME_LENGTH = 20;
var TEST_ACTUAL_LENGTH = 37;
var TEST_EXPECTED_LENGTH = 37;
var TAB = "  ";
var ConsolePrinter = (function () {
    function ConsolePrinter() {
        this.totalTests = 0;
        this.totalSequences = 0;
        this.totalInteractions = 0;
        this.successfulTests = 0;
        this.successfulSequences = 0;
        this.successfulInteractions = 0;
    }
    ConsolePrinter.rpad = function (s, padString, length) {
        if (!s) {
            s = "";
        }
        if (Array.isArray(s)) {
            s = s.toString();
        }
        while (s.length < length) {
            s = s + padString;
        }
        return s.substring(0, length);
    };
    ConsolePrinter.concat = function (buffer, line, error) {
        var color = chalk.default.green;
        if (error) {
            color = chalk.default.red;
        }
        buffer += color(line) + "\n";
        return buffer;
    };
    ConsolePrinter.prototype.printResultsByFile = function (resultsByFile) {
        var output = "";
        var firstFile = true;
        for (var _i = 0, _a = Object.keys(resultsByFile); _i < _a.length; _i++) {
            var file = _a[_i];
            var result = resultsByFile[file];
            var shortFile = path.basename(file);
            if (firstFile) {
                firstFile = false;
            }
            else {
                output += "\n";
            }
            output += this.printResult(shortFile, result, false);
        }
        output = this.printSummary(output);
        return output;
    };
    ConsolePrinter.prototype.printResult = function (name, result, summarize) {
        if (summarize === void 0) { summarize = true; }
        this.totalTests++;
        var out = chalk.default.green;
        if (result.result !== "success") {
            out = chalk.default.red;
        }
        else {
            this.successfulTests++;
        }
        var line = name + "\n";
        var output = out(line);
        var firstSequence = true;
        for (var _i = 0, _a = result.tests; _i < _a.length; _i++) {
            var testResult = _a[_i];
            this.totalInteractions++;
            if (testResult.test.sequenceIndex === 1) {
                this.totalSequences++;
                var sequenceHasError = this.sequenceHasError(testResult.test.sequence, result.tests);
                if (!sequenceHasError) {
                    this.successfulSequences++;
                }
                if (firstSequence) {
                    firstSequence = false;
                }
                else {
                    output += "\n";
                }
                var sequenceLine = TAB + "Sequence " + testResult.test.sequence + ": " + testResult.test.input;
                output = ConsolePrinter.concat(output, sequenceLine, sequenceHasError);
            }
            line = TAB + TAB;
            if (testResult.result !== "success") {
                var error = testResult.errors[0];
                line += ConsolePrinter.rpad(testResult.test.input, " ", TEST_NAME_LENGTH)
                    + "  " + error.property;
                var actualLine = TAB + TAB + TAB + "Actual:   " + error.actual;
                var expectedLine = TAB + TAB + TAB + "Expected: " + error.expected;
                output = ConsolePrinter.concat(output, line, true);
                output = ConsolePrinter.concat(output, actualLine, true);
                output = ConsolePrinter.concat(output, expectedLine, true);
            }
            else {
                this.successfulInteractions++;
                var actual = testResult.actual ? testResult.actual.transcript : "";
                var expected = testResult.test.expected ? testResult.test.expected.transcript : "";
                line += ConsolePrinter.rpad(testResult.test.input, " ", TEST_NAME_LENGTH)
                    + "  Actual: " + ConsolePrinter.rpad(actual, " ", TEST_ACTUAL_LENGTH)
                    + "  Expected: " + ConsolePrinter.rpad(expected, " ", TEST_EXPECTED_LENGTH);
                output = ConsolePrinter.concat(output, line, false);
            }
        }
        if (summarize) {
            output = this.printSummary(output);
        }
        return output;
    };
    ConsolePrinter.prototype.printSummary = function (output) {
        var summaryPadding = 3;
        output += "\n";
        var hasErrors = (this.totalTests !== this.successfulTests);
        output = ConsolePrinter.concat(output, "Summary:", hasErrors);
        output = ConsolePrinter.concat(output, "Files:        "
            + ConsolePrinter.rpad(this.totalTests + "", " ", summaryPadding)
            + " Successful: " + ConsolePrinter.rpad(this.successfulTests + "", " ", summaryPadding)
            + " Failed: " + (this.totalTests - this.successfulTests), hasErrors);
        output = ConsolePrinter.concat(output, "Sequences:    " + ConsolePrinter.rpad(this.totalSequences + "", " ", summaryPadding)
            + " Successful: " + ConsolePrinter.rpad(this.successfulSequences + "", " ", summaryPadding)
            + " Failed: " + (this.totalSequences - this.successfulSequences), hasErrors);
        output = ConsolePrinter.concat(output, "Interactions: " + ConsolePrinter.rpad(this.totalInteractions + "", " ", summaryPadding)
            + " Successful: " + ConsolePrinter.rpad(this.successfulInteractions + "", " ", summaryPadding)
            + " Failed: " + (this.totalInteractions - this.successfulInteractions), hasErrors);
        return output;
    };
    ConsolePrinter.prototype.sequenceHasError = function (sequence, testResults) {
        for (var _i = 0, testResults_1 = testResults; _i < testResults_1.length; _i++) {
            var testResult = testResults_1[_i];
            if (testResult.test.sequence === sequence && testResult.result !== "success") {
                return true;
            }
        }
        return false;
    };
    return ConsolePrinter;
}());
exports.ConsolePrinter = ConsolePrinter;
