// virtual-device-sdk/lib/src/YAMLParser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var YAMLParser = (function () {
    function YAMLParser(yamlContents) {
        this.yamlContents = yamlContents;
    }
    YAMLParser.countAtStart = function (line, value) {
        var count = 0;
        while (true) {
            if (line.startsWith(value)) {
                count++;
                line = line.slice(value.length);
            }
            else {
                break;
            }
        }
        return count;
    };
    YAMLParser.prototype.parse = function () {
        var lines = this.yamlContents.split(YAMLParser.NEWLINE);
        var context = new YAMLContext();
        context.push(new Value(0, 0, undefined, []));
        for (var _i = 0, lines_1 = lines; _i < lines_1.length; _i++) {
            var line = lines_1[_i];
            if (line.startsWith("#")) {
                continue;
            }
            context.lineNumber++;
            this.parseLine(line, context);
        }
        return context.root().array();
    };
    YAMLParser.prototype.parseLine = function (line, context) {
        var cleanLine = line.trim();
        var tabs = 0;
        if (cleanLine.length > 0) {
            YAMLParser.countAtStart(line, YAMLParser.TAB);
            if (tabs === 0) {
                tabs = YAMLParser.countAtStart(line, YAMLParser.TWO_SPACES);
            }
        }
        if (tabs - context.tabs > 1) {
            throw new Error("INVALID - Line " + context.lineNumber + ": previous line is "
                + context.tabs + ". This line is: " + tabs
                + ". Cannot increase indent by more than one tab");
        }
        context.popTo(tabs);
        if (cleanLine.length === 0) {
            context.push(new Value(context.lineNumber, tabs, undefined, null));
        }
        else if (cleanLine.endsWith(":")) {
            var name = cleanLine.substring(0, cleanLine.length - 1);
            context.push(new Value(context.lineNumber, tabs, name));
        }
        else if (cleanLine.startsWith("- ")) {
            context.top().value([]);
            var arrayValue = cleanLine.substring(2).trim();
            context.top().array().push(new Value(context.lineNumber, tabs, undefined, arrayValue));
        }
        else {
            var name = cleanLine.split(":")[0].trim();
            var value = cleanLine.split(":").slice(1).join(":").trim();
            context.push(new Value(context.lineNumber, tabs, name, value));
        }
    };
    YAMLParser.TAB = "\t";
    YAMLParser.NEWLINE = "\n";
    YAMLParser.TWO_SPACES = "  ";
    return YAMLParser;
}());
exports.YAMLParser = YAMLParser;
var YAMLContext = (function () {
    function YAMLContext() {
        this.lineNumber = 0;
        this.tabs = 0;
        this.stack = [];
    }
    YAMLContext.prototype.top = function () {
        return this.stack[this.stack.length - 1];
    };
    YAMLContext.prototype.root = function () {
        return this.stack[0];
    };
    YAMLContext.prototype.popTo = function (tabs) {
        this.tabs = tabs;
        if (tabs >= this.stack.length) {
            return;
        }
        this.stack = this.stack.slice(0, tabs + 1);
    };
    YAMLContext.prototype.push = function (value) {
        if (this.stack.length > 0) {
            if (this.top().isArray()) {
                this.top().array().push(value);
            }
            else {
                if (!this.top().value()) {
                    this.top().value({});
                }
                this.top().value()[value.name()] = value;
            }
        }
        this.stack.push(value);
    };
    return YAMLContext;
}());
exports.YAMLContext = YAMLContext;
var Value = (function () {
    function Value(line, tabs, _name, _value) {
        this.line = line;
        this.tabs = tabs;
        this._name = _name;
        this._value = _value;
    }
    Value.cleanString = function (s) {
        if (!s) {
            return undefined;
        }
        if (s.startsWith("'") || s.startsWith("\"")) {
            s = s.substring(1);
        }
        if (s.endsWith("'") || s.endsWith("\"")) {
            s = s.substring(0, s.length - 1);
        }
        return s;
    };
    Value.prototype.name = function () {
        return Value.cleanString(this._name);
    };
    Value.prototype.isArray = function () {
        return Array.isArray(this._value);
    };
    Value.prototype.isNull = function () {
        return this._value === null;
    };
    Value.prototype.isString = function () {
        return (typeof this._value) === "string";
    };
    Value.prototype.isEmpty = function () {
        return this._value.length === 0;
    };
    Value.prototype.isObject = function () {
        return (typeof this._value) === "object";
    };
    Value.prototype.object = function () {
        var o = this._value;
        var newObject = {};
        for (var _i = 0, _a = Object.keys(o); _i < _a.length; _i++) {
            var key = _a[_i];
            var value = o[key];
            var flatValue = void 0;
            if (value.isString()) {
                flatValue = value.string();
            }
            else if (value.isArray()) {
                flatValue = value.stringArray();
            }
            else if (value.isObject()) {
                flatValue = value.object();
            }
            newObject[key] = flatValue;
        }
        return newObject;
    };
    Value.prototype.array = function () {
        return this._value;
    };
    Value.prototype.stringArray = function () {
        var strings = [];
        for (var _i = 0, _a = this.array(); _i < _a.length; _i++) {
            var value = _a[_i];
            strings.push(value.string());
        }
        return strings;
    };
    Value.prototype.string = function () {
        return Value.cleanString(this._value);
    };
    Value.prototype.toString = function () {
        return "Name: " + this._name + " Value: " + this._value;
    };
    Value.prototype.value = function (v) {
        if (!v) {
            return this._value;
        }
        if (this._value) {
            return this._value;
        }
        this._value = v;
        return this._value;
    };
    return Value;
}());
exports.Value = Value;
